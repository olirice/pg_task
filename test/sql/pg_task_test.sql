-- Create pg_task 
create schema ext;
create extension pg_task with schema ext;


-- default queue exists
select * from ext.queue;

-- upsert_queue
begin;
    select ext.upsert_queue('heavy', 'FIFO') > 0;
rollback;

-- upsert_queue idempotent
begin;
    select ext.upsert_queue('heavy', 'FIFO') > 0;
    select ext.upsert_queue('heavy', 'FIFO') > 0;
    -- We made one queue, and there is one default queue
    select count(1) = 2 from ext.queue;
rollback;


-- upsert_task_definition
begin;
    select ext.upsert_task_definition('resize_image' ) > 0;
    select ext.upsert_task_definition('resize_image', 'default') > 0;
    select ext.upsert_task_definition(
        'resize_image',
        'default',
        3,
        'CONSTANT',
        '1 minute'
    ) > 0;
rollback;
