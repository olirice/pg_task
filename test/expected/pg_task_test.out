-- Create pg_task 
create schema ext;
create extension pg_task with schema ext;
-- default queue exists
select count(1) from ext.queue;
 count 
-------
     1
(1 row)

-- upsert_queue
begin;
    select ext.upsert_queue('heavy', 'FIFO') > 0;
 ?column? 
----------
 t
(1 row)

rollback;
-- upsert_queue idempotent
begin;
    select ext.upsert_queue('heavy', 'FIFO') > 0;
 ?column? 
----------
 t
(1 row)

    select ext.upsert_queue('heavy', 'FIFO') > 0;
 ?column? 
----------
 t
(1 row)

    -- We made one queue, and there is one default queue
    select count(1) = 2 from ext.queue;
 ?column? 
----------
 t
(1 row)

rollback;
-- upsert_task_definition
begin;
    select ext.upsert_task_definition('resize_image' ) > 0;
 ?column? 
----------
 t
(1 row)

    select ext.upsert_task_definition('resize_image', 'default') > 0;
 ?column? 
----------
 
(1 row)

    select ext.upsert_task_definition(
        'resize_image',
        'default',
        3,
        'CONSTANT',
        '1 minute'
    ) > 0;
 ?column? 
----------
 
(1 row)

rollback;
-- enqueue and get task
begin;
    select ext.enqueue_task('resize_image', '{"height": 10}', timezone('utc', now()));
 enqueue_task 
--------------
            1
(1 row)

    select ext.acquire_task('default') = ('resize_image', '{"height": 10}')::ext.acquired_task;
 ?column? 
----------
 t
(1 row)

    -- No tasks reamin
    select ext.acquire_task('default') is not null;
 ?column? 
----------
 t
(1 row)

rollback;
-- enqueue fail for queue does not exist
begin;
    -- Expected to fail
    select ext.enqueue_task('resize_image', '{}', timezone('utc', now()), 'other-DNE');
ERROR:  queue other-DNEs does not exist
CONTEXT:  PL/pgSQL function ext.enqueue_task(text,jsonb,timestamp without time zone,text) line 39 at RAISE
rollback;
